lib/inky/inky.ex

```diff
@@ -0,0 +1,625 @@
+defmodule Inky do
+  @moduledoc """
+  Documentation for Inky.
+  """
+
+  @doc """
+  Hello world.
+
+  ## Examples
+
+      iex> alias Inky
+      iex> state = Inky.setup(nil, :phat, :red)
+      iex> state = Enum.reduce(0..(state.height - 1), state, fn y, state ->Enum.reduce(0..(state.width - 1), state, fn x, state ->Inky.set_pixel(state, x, y, state.red)end)end)
+      iex> Inky.show(state)
+
+  """
+
+  alias Circuits.SPI
+  alias Circuits.GPIO
+  alias Inky.InkyPhat
+  alias Inky.InkyWhat
+  alias Inky.State
+  # alias Inky.Pixel
+
+  @reset_pin 27
+  @busy_pin 17
+  @dc_pin 22
+
+  # @mosi_pin 10
+  # @sclk_pin 11
+  @cs0_pin 0
+
+  @spi_chunk_size 4096
+  @spi_command 0
+  @spi_data 1
+
+  # Used in logo example
+  # inkyphat and inkywhat classes
+  # color constants: RED, BLACK, WHITE
+  # dimension constants: WIDTH, HEIGHT
+  # PIL: putpixel(value)
+  # set_image
+  # show
+
+  # SPI bus options include:
+  # * `mode`: This specifies the clock polarity and phase to use. (0)
+  # * `bits_per_word`: bits per word on the bus (8)
+  # * `speed_hz`: bus speed (1000000)
+  # * `delay_us`: delay between transaction (10)
+
+  def setup(state \\ nil, type, luts_color)
+      when type in [:phat, :what] and luts_color in [:black, :red, :yellow] do
+    state =
+      case state do
+        %State{} ->
+          state
+
+        nil ->
+          {:ok, dc_pid} = GPIO.open(@dc_pin, :output)
+          {:ok, reset_pid} = GPIO.open(@reset_pin, :output)
+          {:ok, busy_pid} = GPIO.open(@busy_pin, :input)
+          # GPIO.write(gpio_pid, 1)
+          {:ok, spi_pid} = SPI.open("spidev0." <> to_string(@cs0_pin), speed_hz: 488_000)
+          # Use binary pattern matching to pull out the ADC counts (low 10 bits)
+          # <<_::size(6), counts::size(10)>> = SPI.transfer(spi_pid, <<0x78, 0x00>>)
+          %State{
+            dc_pid: dc_pid,
+            reset_pid: reset_pid,
+            busy_pid: busy_pid,
+            spi_pid: spi_pid,
+            color: luts_color
+          }
+      end
+
+    GPIO.write(state.reset_pid, 0)
+    :timer.sleep(100)
+    GPIO.write(state.reset_pid, 1)
+    :timer.sleep(100)
+
+    state =
+      case type do
+        :phat -> InkyPhat.update_state(state)
+        :what -> InkyWhat.update_state(state)
+      end
+
+    soft_reset(state)
+    busy_wait(state)
+    state
+  end
```

Comment:
With all the suggestions above you could transform this function into a nice elixir-style pipeline:

```elixir
state
|> prepare_state()
|> hard_reset() # Here you can do the reset pin writing, make sure to return the state
|> update_state(for: type)
|> soft_reset()
|> busy_wait()
```

When I look at this you could most likely also consider renaming a few functions to make it more "prose"-/"recipe"-like. Because this doesn't really tell you what it does.

Reply:
Oooh, I like this refactor. And I get what you mean, it could be more descriptive. When I wrote it I couldn't have told you what half of it did. Now I probably can :)